### ISSUE TITLE
Extracting layers with negative canvas offsets from PSD translates offsets to large positive values

### ISSUE THREAD

**Issue by KurtPfeifle** [ImageMagick/ImageMagick#1061]
I've tested this with today's Git checkout of sources which I built myself on macOS. Unfortunately I cannot share the PSD file (it's not mine, and I'm not allowed to share it by the owner).

````
/usr/local/bin/magick -version
 Version: ImageMagick 7.0.7-29 Q16 x86_64 2018-03-26 http://www.imagemagick.org
 Copyright: © 1999-2018 ImageMagick Studio LLC
 License: http://www.imagemagick.org/script/license.php
 Features: Cipher DPC HDRI Modules 
 Delegates (built-in): autotrace bzlib cairo djvu fftw fontconfig freetype gslib gvc jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png ps raw rsvg tiff webp wmf x xml zlib
````

To see the problem, extract a layer with negative canvas offsets (here layer 1):

````
magick in.psd[1] layer1.png

/usr/local/bin/magick identify   in.psd[1] layer1.png
   in.psd[1]=>in.psd PSD 2433x2442 2433x2442-584-642 8-bit sRGB 0.000u 0:00.000
   layer1.png PNG 2433x2442 2433x2442+4294966712+4294966654 8-bit sRGB 420877B 0.000u 0:00.000
````

The formula it translates negative offsets into positive ones seems to be this:

      newX = X + 4 * (2^30)
      newY = Y + 4 * (2^30)

that is, it adds 4.294.967.296 to the original negative offset.

That probably wouldn't be a major problem in itself -- but ImageMagick isn't able to further process a file with the new offset values, for instance re-combining such extracted layers with other layers from the original PSD file.

**Comment by urban-warrior** [comment]
Unfortunately, without an example PSD image that exhibits this problem, there is little we can do.  We did check the PSD module and it is reading the offsets as signed integers which permit negative offsets.  Add `-debug coder` to your command-line and post the results here.  The debugging may help us debug the problem.

**Comment by KurtPfeifle** [comment]
I found an example Photoshop file with one layer using a negative offset, downloadable here as a ZIP:

* https://www.freepik.com/free-psd/wooden-button-design_989771.htm#term=buttons&page=1&position=1

It's under a "free license with attribution": Created by Teinstud - Freepik.com.

The issue reported here can be reproduced with the PSD file included in the downloaded .zip.

**Comment by dlemstra** [comment]
I cannot reproduce your issue with the test file:
```
I:\psd>magick convert OHIGEM1.psd[2] layer2.png

I:\psd>magick identify OHIGEM1.psd[2] layer2.png
OHIGEM1.psd[2]=>OHIGEM1.psd PSD 3027x2941 3027x2941-100+620 8-bit sRGB 0.000u 0:00.000
layer2.png PNG 3027x2941 3027x2941-100+620 8-bit sRGB 1.88526MiB 0.000u 0:00.000
```

Which version of libpng are you using?

**Comment by KurtPfeifle** [comment]
I saw this problem on macOS. I'm not sure how to see which libpng is used here:

    otool -L /usr/local/bin/magick | grep -i png

returns an empty line...  With the MacPorts-installed libpng I've gotten:

    /opt/local/bin/libpng-config --version
     1.6.34

I'll quickly build the latest on Debian Stretch, see if the issue is reproducable there too (or not) and report back in a few minutes...

**Comment by fmw42** [comment]
To see which version of libpng is being used, you can do 

`convert -list format`

and look at the end of the line for PNG

**Comment by KurtPfeifle** [comment]
@fmw42 👍 

Thanks, Fred. After a while, I also remembered these two commands. But (of course) they didn't give a different result than what `'libpng-config --version'` had reported.

**Comment by KurtPfeifle** [comment]
To investigate a bit more, I ran the same tests on...

* two Raspberry Pi systems (with *openSUSE Tumbleweed* and *Raspbian Jessie*) as well as on
* two x64 CPUs (with *Debian Stretch* and *macOS*),

...each time with a recently locally built IM from Git sources.

## TL;DR :

On openSUSE Tumbleweed **with libpng 1.2.50, the problem does not occur.**    
On 3 other systems **with libpng versions 1.6.28, 1.6.31/1.6.34 this issue is present.**

----

Here are more details about my tests:

## macOS

1. **libpng version** ( = 1.6.34)

    `````
    /usr/local/bin/magick -list format | grep "libpng "
      MNG* PNG       rw+   Multiple-image Network Graphics (libpng 1.6.34)
      PNG* PNG       rw-   Portable Network Graphics (libpng 1.6.34)
    `````

1. **magick version**

    ````
     /usr/local/bin/magick -version
     Version: ImageMagick 7.0.7-29 Q16 x86_64 2018-03-26 http://www.imagemagick.org
     Copyright: © 1999-2018 ImageMagick Studio LLC
     License: http://www.imagemagick.org/script/license.php
     Features: Cipher DPC HDRI Modules 
     Delegates (built-in): autotrace bzlib cairo djvu fftw fontconfig freetype gslib gvc \
         jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png ps raw rsvg tiff webp \
         wmf x xml zlib
    ````

1. **reproducing this issue:** ***YES***

    ````
     /usr/local/bin/magick OHIGEM1.psd[2] l2.png
     /usr/local/bin/magick identify -format "%g\n" l2.png
      3027x2941+4294967196+620
    ````

## Debian Stretch (x64)

1. **libpng version** ( = 1.6.28)

    `````
    /usr/local/bin/magick -list format | grep "libpng "
      MNG* rw+   Multiple-image Network Graphics (libpng 1.6.28)
      PNG* rw-   Portable Network Graphics (libpng 1.6.28)
    `````

1. **magick version**

    ````
    /usr/local/bin/magick -version
     Version: ImageMagick 7.0.7-29 Q16 x86_64 20180324 http://www.imagemagick.org
     Copyright: © 1999-2018 ImageMagick Studio LLC
     License: http://www.imagemagick.org/script/license.php
     Features: Cipher DPC HDRI OpenMP 
     Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib gvc heic jbig \
         jng jp2 jpeg lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp wmf \
         x xml zlib
    ````

1. **reproducing this issue:** ***YES***

    ````
     /usr/local/bin/magick OHIGEM1.psd[2] l2.png
     /usr/local/bin/magick identify -format "%g\n" l2.png
      3027x2941+4294967196+620
    ````


## Raspbian (Debian Jessie) on Raspberry Pi 2

1. **libpng version** ( = 1.6.31 + 1.6.34 ??)

    ````
    /usr/local/bin/magick -list format | grep "libpng "
      MNG* PNG       rw+   Multiple-image Network Graphics (libpng 1.6.31,1.6.34)
      PNG* PNG       rw-   Portable Network Graphics (libpng 1.6.31,1.6.34)
    ````

1. **magick version**

    ````
    /usr/local/bin/magick -version
     Version: ImageMagick 7.0.7-25 Q16 aarch64 2018-02-27 http://www.imagemagick.org
     Copyright: © 1999-2018 ImageMagick Studio LLC
     License: http://www.imagemagick.org/script/license.php
     Features: Cipher DPC HDRI Modules OpenMP 
     Delegates (built-in): autotrace bzlib cairo djvu fftw flif fontconfig freetype gvc \
         jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png raw rsvg tiff \
         webp wmf x xml zlib
    ````

1. **reproducing this issue:** ***YES***

    ````
     /usr/local/bin/magick OHIGEM1.psd[2] l2.png
     /usr/local/bin/magick identify -format "%g\n" l2.png
      3027x2941+4294967196+620
    ````

## openSUSE Tumbleweed on Raspberry 2

1. **libpng version** ( = 1.2.50)

    ````
    /usr/local/bin/magick -list format | grep "libpng "
      MNG* PNG       rw+   Multiple-image Network Graphics (libpng 1.2.50)
      PNG* PNG       rw-   Portable Network Graphics (libpng 1.2.50)
    ````

1. **magick version**

    ````
    /usr/local/bin/magick -version
     Version: ImageMagick 7.0.7-28 Q16 armv7l 2018-03-24 http://www.imagemagick.org
     Copyright: © 1999-2018 ImageMagick Studio LLC
     License: http://www.imagemagick.org/script/license.php
     Features: Cipher DPC HDRI Modules OpenMP 
     Delegates (built-in): autotrace bzlib cairo djvu fftw fontconfig freetype gvc jbig jng \
          jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png rsvg tiff webp wmf x xml zlib
    ````

1. **reproducing this issue:** ***NO***

    ````
     /usr/local/bin/magick OHIGEM1.psd[2] l2.png
     /usr/local/bin/magick identify -format "%g\n" l2.png
      3027x2941-100+620
    ````

**Comment by KurtPfeifle** [comment]
@dlemstra -- out of curiosity: which was the version of libpng you used when you could not reproduce the issue?

**Comment by dlemstra** [comment]
`1.6.34` on Windows.
